File: fs/Practice2/DynamicProgramming/Section/Cake2.fs
// https://atcoder.jp/contests/joi2015ho/tasks/joi2015ho_b

stdin.ReadLine ()
|> int
|> fun n ->
  (
    (Array.init n (fun _ -> stdin.ReadLine () |> int64) |> fun arr -> [ arr; arr ] |> Array.concat),
    (Array.init n (fun _ -> Array.create n 0L)),
    [| fun _ _ -> 0L |]
  )
  |> fun (cakes, memo, fnc) ->
    (
      (* "start" and "count" mean the range of this turn *)
      fun start count ->
        match count with
        | 0 -> cakes.[start]
        | x when x < 0 -> 0L
        (* If the memo has a value, return it *)
        | _ when memo.[start % n].[count] <> 0L -> memo.[start % n].[count]
        | _ ->
          [
            (start,         [(start + 1, 2); (start + count, 1)]); (* JOI takes a value from head *)
            (start + count, [(start, 1); (start + count - 1, 0)]); (* JOI takes a value from last *)
          ]
          |> List.map (
            fun (joi, ioi) ->
              ioi
              (* IOI takes a larger value *)
              |> List.maxBy (fun (i, _) -> cakes.[i])
              (* Getting a value from the rest of the range *)
              |> fun (_, offset) -> (fnc.[0] (start + offset) (count - 2))
              (* Adding the value acquired this turn *)
              |> (+) cakes.[joi]
          )
          (* Adopting the maximum value of this range *)
          |> List.max
          (* Setting the value in the memo and return it *)
          |> fun v -> (memo.[start % n].[count] <- v) |> fun () -> v
    )
    |> fun fn ->
      (fnc.[0] <- fn)
      |> fun () ->
        [0..(n - 1)]
        |> List.map (fun i -> fn i (n - 1))
        |> List.max
|> printfn "%d"

// https://atcoder.jp/contests/joi2015ho/submissions/20004942
// https://atcoder.jp/contests/joi2015ho/submissions/20005189

File: scala/Practice2/DynamicProgramming/Section/Cake2.scala
// https://atcoder.jp/contests/joi2015ho/tasks/joi2015ho_b

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    io.StdIn.readLine.toInt
    |> ((n) =>
      (
        ((1 to n).map(_x => io.StdIn.readLine.toLong).toArray |> (arr => Array.concat(arr, arr))),
        ((1 to n).map(_x => Array.fill[Long](n)(0))).toArray,
        Array((_s: Int, _c: Int) => 0L)
      )
      |> { case (cakes, memo, fnc) => (
        ((start: Int, count: Int) => (
          count match {
            case 0 => cakes(start)
            case x if x < 0 => 0L
            case _ if memo(start % n)(count) != 0L => memo(start % n)(count)
            case _ => (
              List(
                (start,         List((start + 1, 2), (start + count, 1))),
                (start + count, List((start, 1), (start + count - 1, 0))),
              )
              .map {
                case (joi, ioi) => (
                  ioi.maxBy { case (i, _x) => cakes(i) }
                  |> { case (_x, offset) => fnc(0)((start + offset), (count - 2)) }
                  |> (_ + cakes(joi))
                )
              }
              .max
              |> (v => (memo(start % n)(count) = v) |> (_ => v))
            )
          }
        ))
        |> (fn => (
          (fnc(0) = fn)
          |> (_ => (
            (0 until n)
              .map(i => fn(i, (n - 1)))
              .max
          ))
        ))
      )}
    )
    |> println
  )
}

// https://atcoder.jp/contests/joi2015ho/submissions/20033778
